### PR TITLE
Resolve the GrainType from the provided generic grain value

### LINKED ISSUE #140: GrainType on GrainId is duplicate of GrainIdentity
The GrainType value of the GrainId is the same value as the grain key so if you're expecting a type you'd get an incorrect value.

Orleans resolves this value via the grain interface although you can use the [GrainTypeResolver](https://github.com/dotnet/orleans/blob/main/src/Orleans.Core/Manifest/GrainTypeResolver.cs) to take the Type of the grain and parse the GrainType correctly.

I've submitted a PR with a fix, since there's a dependency tree to construct a `GrainTypeResolver` I've been lazy and used a DI container to correctly resolve this using the Orleans methods since some of the dependencies are internal to Orleans. This could probably be cached but i noticed all dependencies in TestKit are rooted to the silo instance so kept this the same.

### PR DESCRIPTION
**Pull request by Romanx**
closes #140

### CODE REVIEW

File: src/OrleansTestKit/TestKitSilo.cs [hunk 1]
```diff
@@ -41,6 +47,14 @@ public TestKitSilo()
...
         ServiceProvider.AddService<IGrainRuntime>(_grainRuntime);
         _grainCreator = new TestGrainCreator(_grainRuntime, ServiceProvider);
+
+        var provider = new ServiceCollection()
+            .AddSingleton<GrainTypeResolver>()
+            .AddSingleton<IGrainTypeProvider, AttributeGrainTypeProvider>()
+            .AddSerializer()
+            .BuildServiceProvider();
+
+        _grainTypeResolver = provider.GetRequiredService<GrainTypeResolver>();
```
Comment (seniorquico) [inline src/OrleansTestKit/TestKitSilo.cs:57]: Why do we need a DI container just to create this instance?

**Review by seniorquico — COMMENTED** [review]
I'm sorry, I don't follow what problem this solves (I don't quite get the description in #140). Can you please add a test case to this PR that demonstrates the fix?

**Comment by Romanx** [comment]
Hey @seniorquico,

Sorry it was late on Friday and didn't give a good description. This corrects the GrainType on the created grains to be the same as Orleans so anything that is looking at the Type value, including any aliases, are correct. I can add a test for that specific case.

We had some code that looks the GrainType and was failing since it was an invalid value.

I used a DI container since the `GrainTypeResolver` has an interesting dependency tree and this approach handles Orleans changing how this value is generated since it's extensible. I could make it more explicit but there might have to be manual changes if Orleans changes functionality. 

I'd rather not be creating the container and resolving the value each Silo but TestKit seems to isolate itself on that level so it seemed sensible. It could be made private and static so it's only resolved since it shouldn't change but didn't want to assume.

**Comment by Romanx** [comment]
@seniorquico Sorry for the delay, added tests showing what this change provides.

**Comment by Romanx** [comment]
@seniorquico Any thoughts on this? It's a blocker for using testkit due to the GrainType and I can't really use reflection to hack around it since most of the other pieces in CreateGrainAsync are internal so i can't duplicate it locally.

**Comment by Romanx** [comment]
@seniorquico Another gentle nudge on this since all our tests are broken without this change and theres no way for me to hack around it so it's blocking us. Happy to make more changes if you'd like

**Comment by Romanx** [comment]
@seniorquico Another bump on this. Happy to change approach with some feedback. 

Looked at the latest changes with grain context. It gets me some way to being able to do this externally to testkit but even if I override the `GrainContext` with the "correct" grain id with the grain type correctly, the context gets replaced since the comparison is done on the whole grain id.

If you'd rather not the service provider route so the grain types are correct for all grains I could instead expose an overload of `GetOrAddGrainContext` that takes a whole GrainId and allows you to set that?

File: src/OrleansTestKit/TestKitSilo.cs (hunk 1 above)
Comment (Romanx) [reply src/OrleansTestKit/TestKitSilo.cs:61]: Commented on the PR with a more full description.

**Comment by fergusbown** [comment]
HI, sorry to keep prodding, I work with @Romanx - any chance of progressing this, we could really do with it :)  Thanks!

**Review by seniorquico — APPROVED** [review]

**Pull request merged**